### PR TITLE
Support Output Format

### LINKED ISSUE #28: Implement OutputFormat

### PR DESCRIPTION
**Pull request by yangb8**
Closes #28

### CODE REVIEW

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (vijikarthi) [inline src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: typo - change input to output

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (vijikarthi) [inline src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: typo - change input to output

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (vijikarthi) [inline src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java:50]: is it not require to validate if the output stream is already available and populated with any data?

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (vijikarthi) [inline src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: typo - change input to output

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (vijikarthi) [inline src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: typo - change input to output

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputRecordWriter.java
Comment (vijikarthi) [inline src/main/java/io/pravega/connectors/hadoop/PravegaOutputRecordWriter.java:105]: will the writer and executor service will not be closed in case of an error?

**Review by vijikarthi — CHANGES_REQUESTED** [review]

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (yangb8) [reply src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: accepted

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (yangb8) [reply src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: accepted

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (yangb8) [reply src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java:50]: it's a valid case, the user can append events to an existing stream.

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (yangb8) [reply src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: accepted

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (yangb8) [reply src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: accepted

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputRecordWriter.java
Comment (yangb8) [reply src/main/java/io/pravega/connectors/hadoop/PravegaOutputRecordWriter.java:105]: accepted. add try-catch block to make sure they can be closed even if the previous statement throws an exception.

**Comment by vijikarthi** [comment]
Can we merge this code if there are no feedback to address?

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (EronWright) [inline src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: Please share some documentation or reference code to explain the output committer.   Is it reasonable to use the `/tmp` path, and to use the `jobID` rather than `taskAttemptID`?

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (EronWright) [inline src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: A few cut-and-paste mistakes in this method.   Would also be nice to use a helper method here.

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (EronWright) [inline src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: Couldn't you create the `ClientFactory` in the constructor?   Ugly code otherwise, passing `controllerURI` and `scopeName` for no other reason.

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (EronWright) [inline src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: Let's not create scopes like this anymore.  Apps should throw when the scope doesn't exist.   Consider that an ACL is needed in advance anyway.

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (EronWright) [inline src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: Seems wrong to have `fixed` be the only scaling policy...

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (EronWright) [inline src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: Please ensure that `streamName` supports a fully-qualified stream name, with `scopeName` being an optional **default** scope to be used when the stream name is not qualified.

File: src/main/java/io/pravega/connectors/hadoop/PravegaEventRouter.java
Comment (EronWright) [inline src/main/java/io/pravega/connectors/hadoop/PravegaEventRouter.java:21]: Please mark this interface as `@FunctionalInterface`

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (EronWright) [inline src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: I don't understand why `FileOutputCommitter` would be useful here. 
https://github.com/apache/hadoop/blob/trunk/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/lib/output/FileOutputCommitter.java

**Review by EronWright — DISMISSED** [review]

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (vijikarthi) [reply src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: I think the implementation is modeled based on https://github.com/kafka-dev/kafka/blob/master/contrib/hadoop-producer/src/main/java/kafka/bridge/hadoop/KafkaOutputFormat.java#L70?

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (vijikarthi) [reply src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: modified the implementation to make it generic to retrieve the dynamic class instance

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (vijikarthi) [reply src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: Job configuration is not available during object construction to retrieve the scope/URI details to create the client factory instance

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (vijikarthi) [reply src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: (done)

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (vijikarthi) [reply src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: removed stream create step and expects stream to be created ahead of time.

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (vijikarthi) [reply src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: (done)

File: src/main/java/io/pravega/connectors/hadoop/PravegaEventRouter.java
Comment (vijikarthi) [reply src/main/java/io/pravega/connectors/hadoop/PravegaEventRouter.java:21]: (done)

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (vijikarthi) [reply src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: My understanding is that the output committer is used to keep track of the task progress. Do you see a problem in using `FileOutputCommitter`?

**Review by vijikarthi — APPROVED** [review]

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormatBuilder.java
Comment (fpj) [inline src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormatBuilder.java:67]: Can you fix the javadocs in this class, please?

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputRecordWriter.java
Comment (fpj) [inline src/main/java/io/pravega/connectors/hadoop/PravegaOutputRecordWriter.java]: can we simplify these two lines by doing something like:

```
final CompletableFuture<Void> future = 
                        (pravegaEventRouter == null) ? writer.writeEvent(value); :
                                                       writer.writeEvent(pravegaEventRouter.getRoutingKey(value), value);
```

File: README.md
Comment (fpj) [inline README.md]: Replace the first sentence with:

```
Implements both the input and the output format interfaces for Hadoop.
```

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormatBuilder.java
Comment (fpj) [inline src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormatBuilder.java]: I'm not sure what this is doing. It looks like it is trying to configure the number of segments for a fixed stream, but it is not used anywhere. It is also not clear what happens when the stream auto-scales. Maybe remove it?

**Review by fpj — CHANGES_REQUESTED** [review]
Looks good, I have a few minor points.

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (kevinhan88) [inline src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: This location may conflict with other job easily. Suggest to make it configurable, with default path something like "/tmp/pravega/output/" + taskAttemptContext.getTaskAttemptID().getJobID().toString()).

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputRecordWriter.java
Comment (kevinhan88) [inline src/main/java/io/pravega/connectors/hadoop/PravegaOutputRecordWriter.java]: Is number 5 based on some experience? Please consider make it configurable, or at least a static variable if hard-coding is necessary.

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormatBuilder.java
Comment (vijikarthi) [reply src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormatBuilder.java:67]: (done)

File: README.md
Comment (vijikarthi) [reply README.md]: (done)

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormatBuilder.java
Comment (vijikarthi) [reply src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormatBuilder.java]: Yes, it is unused. I have removed it.

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (vijikarthi) [reply src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: The job id will be unique and I don't see an issue with name conflict.

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputRecordWriter.java
Comment (vijikarthi) [reply src/main/java/io/pravega/connectors/hadoop/PravegaOutputRecordWriter.java]: Moved the hard coded value to a variable

File: src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java
Comment (kevinhan88) [reply src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: Yes job id is unique, but hadoop built-in functions for the job may dump temp files into the same location? Just a safer approach though, I'm fine for what it is now.

Comment (vijikarthi) [reply src/main/java/io/pravega/connectors/hadoop/PravegaOutputFormat.java]: (done)

**Review by fpj — APPROVED** [review]

**Pull request merged**
